File: python/filters.py
import logging

from openff.qcsubmit.results import (
    OptimizationResultCollection,
    TorsionDriveResultCollection,
)
from openff.qcsubmit.results.filters import (
    ConformerRMSDFilter,
    ConnectivityFilter,
    ElementFilter,
    RecordStatusFilter,
    SinglepointRecordFilter,
    UnperceivableStereoFilter,
)
from openff.toolkit.utils.exceptions import (
    ChargeCalculationError,
    ConformerGenerationError,
)
from openff.toolkit.utils.toolkits import OpenEyeToolkitWrapper
from qcportal.record_models import RecordStatusEnum

logging.getLogger("openff").setLevel(logging.ERROR)


class ChargeCheckFilter(SinglepointRecordFilter):
    def _filter_function(self, result, record, molecule) -> bool:
        try:
            OpenEyeToolkitWrapper().assign_partial_charges(
                molecule, partial_charge_method="am1bccelf10"
            )
        except (ChargeCalculationError, ConformerGenerationError):
            return False
        else:
            return True


def filter_opt_data(
    dataset: OptimizationResultCollection,
    records_to_remove: list[int],
    include_iodine: bool,
    max_opt_conformers: int,
):
    key = list(dataset.entries.keys())[0]

    dataset.entries[key] = [
        entry
        for entry in dataset.entries[key]
        if entry.record_id not in records_to_remove
    ]

    elements = ["H", "C", "N", "O", "S", "P", "F", "Cl", "Br"]
    if include_iodine:
        elements.append("I")

    return dataset.filter(
        RecordStatusFilter(status=RecordStatusEnum.complete),
        ConnectivityFilter(tolerance=1.2),
        UnperceivableStereoFilter(),
        ElementFilter(allowed_elements=elements),
        ConformerRMSDFilter(max_conformers=max_opt_conformers),
    )


def filter_td_data(
    dataset: TorsionDriveResultCollection,
    records_to_remove: list[int],
    include_iodine: bool = False,
):
    key = list(dataset.entries.keys())[0]

    dataset.entries[key] = [
        entry
        for entry in dataset.entries[key]
        if entry.record_id not in records_to_remove
    ]

    elements = ["H", "C", "N", "O", "S", "P", "F", "Cl", "Br"]
    if include_iodine:
        elements.append("I")

    return dataset.filter(
        RecordStatusFilter(status=RecordStatusEnum.complete),
        ConnectivityFilter(tolerance=1.2),
        ChargeCheckFilter(),
    )


if __name__ == "__main__":
    dataset = OptimizationResultCollection.parse_file(
        "testfiles/download_opt_want.json",
    )
    records_to_remove = [
        2002949,
        2002950,
        18433638,
        18433906,
        2002933,
        2002934,
        2002937,
        2003047,
        2003043,
        95602295,
        95602250,
        18433502,
        18434090,
        2002949,
        2002950,
        18433638,
        18433906,
        2002933,
        2002934,
        2002937,
        2003047,
        2003043,
        95602295,
        95602250,
        18433502,
        18434090,
        18433675,
        18433675,
        2003404,
        2002930,
        2002929,
        2002979,
    ]
    body = filter_opt_data(dataset, records_to_remove, False, 12)
    with open("testfiles/filters_no_charge_want.json", "w") as out:
        out.write(body.json(indent=2))

File: tests/filters.rs
use std::fs::read_to_string;

use openff_qcsubmit::results::{
    filters::{
        ConformerRMSDFilter, ConnectivityFilter, ElementFilter, Filter,
        RecordStatusFilter, UnperceivableStereoFilter,
    },
    BaseResultCollection, OptimizationResultCollection,
};
use qcportal::record_models::RecordStatus;

struct ChargeCheckFilter;

impl Filter<OptimizationResultCollection> for ChargeCheckFilter {
    fn apply(
        &self,
        mut dataset: OptimizationResultCollection,
    ) -> OptimizationResultCollection {
        let entries = dataset.entries();
        dataset.set_entries(entries);
        // TODO update provenance section with applied filters
        dataset
    }
}

fn filter_opt_data(
    mut dataset: OptimizationResultCollection,
    records_to_remove: Vec<usize>,
    include_iodine: bool,
    max_opt_conformers: usize,
) -> OptimizationResultCollection {
    let mut entries = dataset.entries();
    let key = entries.keys().next().unwrap().to_owned();
    let new_entries = entries[&key]
        .iter()
        .filter(|entry| !records_to_remove.contains(&entry.record_id()))
        .cloned()
        .collect();
    entries.insert(key.into(), new_entries);

    let mut elements = vec!["H", "C", "N", "O", "S", "P", "F", "Cl", "Br"];
    if include_iodine {
        elements.push("I");
    }

    dataset.set_entries(entries);
    // filter is going to take a &[Box<dyn Filter>]
    dataset.filter(&[
        Box::new(RecordStatusFilter::new(RecordStatus::Complete)),
        Box::new(ConnectivityFilter::new(1.2)),
        Box::new(UnperceivableStereoFilter::new()),
        Box::new(ElementFilter::new(elements)),
        Box::new(ConformerRMSDFilter::new(max_opt_conformers)),
        Box::new(ChargeCheckFilter),
    ])
}

#[test]
fn filter_opt() {
    // this has 400 entries, I should narrow it down to some subset that still
    // exercises all of the filters
    let dataset = OptimizationResultCollection::parse_file(
        "testfiles/download_opt_want.json",
    )
    .unwrap();
    let records_to_remove = vec![
        2002949, 2002950, 18433638, 18433906, 2002933, 2002934, 2002937,
        2003047, 2003043, 95602295, 95602250, 18433502, 18434090, 2002949,
        2002950, 18433638, 18433906, 2002933, 2002934, 2002937, 2003047,
        2003043, 95602295, 95602250, 18433502, 18434090, 18433675, 18433675,
        2003404, 2002930, 2002929, 2002979,
    ];
    let dataset = filter_opt_data(dataset, records_to_remove, false, 12);
    let got = dataset.json(2);
    let want = read_to_string("testfiles/filters_no_charge_want.json").unwrap();
    assert_eq!(got, want);
}
